File: app/src/main/java/io/github/wulkanowy/ui/base/BasePagerAdapter.java
package io.github.wulkanowy.ui.base;

import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;

import java.util.ArrayList;
import java.util.List;

public class BasePagerAdapter extends FragmentStatePagerAdapter {

    private List<Fragment> fragmentList = new ArrayList<>();

    private List<String> titleList = new ArrayList<>();

    public BasePagerAdapter(FragmentManager fragmentManager) {
        super(fragmentManager);
    }

    public void addFragment(@NonNull Fragment fragment, @NonNull String title) {
        fragmentList.add(fragment);
        titleList.add(title);
    }

    public void addFragment(@NonNull Fragment fragment) {
        fragmentList.add(fragment);
    }

    @Override
    public Fragment getItem(int position) {
        return fragmentList.get(position);
    }

    @Override
    public int getCount() {
        return fragmentList.size();
    }

    @Override
    public CharSequence getPageTitle(int position) {
        if (!titleList.isEmpty()) {
            return titleList.get(position);
        }
        return null;
    }
}

File: app/src/main/java/io/github/wulkanowy/ui/base/BasePagerAdapter.kt
package io.github.wulkanowy.ui.base

import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import android.view.ViewGroup

class BasePagerAdapter(fragmentManager: FragmentManager) : FragmentStatePagerAdapter(fragmentManager) {

    val fragments = mutableMapOf<String?, Fragment>()

    val registeredFragments = mutableMapOf<Int, Fragment>()

    override fun getItem(position: Int) = fragments.values.elementAt(position)

    override fun getCount() = fragments.size

    override fun getPageTitle(position: Int): CharSequence? {
        return fragments.keys.elementAtOrNull(position)
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        return super.instantiateItem(container, position).also {
            registeredFragments[position] = it as Fragment
        }
    }

    override fun destroyItem(container: ViewGroup, position: Int, fragment: Any) {
        registeredFragments.remove(position)
        super.destroyItem(container, position, fragment)
    }
}
